File: Python3/aoc24/day03/solve.py
"""Module containing computation of solutions."""

from .instruction import Do, Dont, Instruction, Mul


def solve_part_one(instructions: list[Instruction]) -> int:
    """Compute solution of part one."""
    return sum(instr.left * instr.right for instr in instructions if isinstance(instr, Mul))


def solve_part_two(instructions: list[Instruction]) -> int:
    """Compute solution of part two."""
    result: int = 0
    enabled: bool = True

    for instr in instructions:
        match instr:
            case Do():
                enabled = True
            case Dont():
                enabled = False
            case Mul(left=left, right=right):
                if enabled:
                    result += left * right

    return result

File: Rust/src/day03/solve.rs
use crate::day03::model::Instruction;
use nom::character::complete::i128;

pub fn solve_part_one(instructions: &[Instruction]) -> i128 {
    let result: i32 = instructions
        .iter()
        .map(|instruction| match instruction {
            Instruction::Mul(left, right) => left * right,
            _ => 0,
        })
        .sum();
    result as i128
}

pub fn solve_part_two(instructions: &[Instruction]) -> i128 {
    let mut enabled: bool = true;
    let mut result: i32 = 0;

    for instr in instructions {
        match instr {
            Instruction::Do => enabled = true,
            Instruction::Dont => enabled = false,
            Instruction::Mul(left, right) => {
                if enabled {
                    result += left * right
                }
            }
        }
    }

    result as i128
}
